File: head_tracking_audio/python_host/dsp/convolution.py
import numpy as np
from scipy.signal import fftconvolve


def fft_convolve_mono_to_stereo(audio, hrir_l, hrir_r):
	if audio.ndim != 1:
		raise ValueError("Audio must be mono (1D)")

	left = fftconvolve(audio, hrir_l, mode="full")
	right = fftconvolve(audio, hrir_r, mode="full")
	return left, right


def normalize_audio(signal):
	peak = np.max(np.abs(signal))
	if peak <= 0.0:
		return signal
	return signal / peak

File: head_tracking_audio/tools/matlab/convolution.m
function [left, right] = fft_convolve_mono_to_stereo(audio, hrir_l, hrir_r)
% FFT_CONVOLVE_MONO_TO_STEREO Convolve mono audio with HRIR left/right
% Returns full convolution (mode='full')
if ndims(audio) ~= 2 && ~isvector(audio)
    error('Audio must be mono vector');
end
audio = audio(:)';
left = conv(audio, hrir_l(:)');
right = conv(audio, hrir_r(:)');
end

function y = normalize_audio(sig)
    peak = max(abs(sig(:)));
    if peak <= 0
        y = sig;
    else
        y = sig / peak;
    end
end
